### ISSUE TITLE
default relative path processing in `include_graphics()` can create broken path

### ISSUE THREAD

**Issue by naikymen** [yihui/knitr#2171]
Hi!

I'm having difficulties including images during knitting.

Here is a reproducible example: [test.zip](https://github.com/yihui/knitr/files/9592044/test.zip)

The notebook is in `R/Rmd1/notebook1.Rmd`. If I run the notebook in RStudio everything works.

Knit the document to get the following error:

![image](https://user-images.githubusercontent.com/3259326/190864969-ba8eeeb5-878a-43a3-968a-c3e6e6bfa82d.png)

Relevant bits about the example:

* I usually keep notebooks in subdirectories of the project, tipically under `R/`.
* I like to set `knitr::opts_knit$set(root.dir = here::here())` at the setup chunk, so paths in chunks are always relative to the project while I'm working.
* I like to save knitted documents to `output/renders` (again relative to the project's root) by setting `knit` in the YAML header to the following function: `rmarkdown::render(inputFile, encoding = encoding, output_dir = "output/renders")`.

If i omit the `include_graphics` call, Knit works and R is able to find the file with `file.exists`:  [notebook1.pdf](https://github.com/yihui/knitr/files/9592072/notebook1.pdf)

If instead of normalizing the path I pass a path relative to the project's root, using `knitr::include_graphics("results/image_1.png")`. I get a different error:

```
! Unable to load picture or PDF file 'results/image_1.png'.
<to be read again> 
                   }
l.174 \includegraphics{results/image_1.png}
```

I would appreciate if someone could point me in the right direction. What path should I pass to `include_graphics` in this case?

I'm submitting this as a bug report because I have been unable to find out why this happens, everything else seems to work, and it is unexpected behavior from my perspective.

Thank you for all the work you've done :)

Best!

N.

---

By filing an issue to this repo, I promise that

- [x] I have fully read the issue guide at https://yihui.org/issue/.
- [x] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('knitr')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('yihui/knitr')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [x] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

Session info:

```
> xfun::session_info("knitr")
R version 4.2.1 (2022-06-23)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Arch Linux, RStudio 2022.2.0.443

Locale:
  LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C               LC_TIME=es_AR.UTF-8       
  LC_COLLATE=en_GB.UTF-8     LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
  LC_PAPER=en_GB.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
  LC_TELEPHONE=C             LC_MEASUREMENT=es_AR.UTF-8 LC_IDENTIFICATION=C       

Package version:
  evaluate_0.16   glue_1.6.2      graphics_4.2.1  grDevices_4.2.1 highr_0.9       knitr_1.40     
  magrittr_2.0.3  methods_4.2.1   stats_4.2.1     stringi_1.7.8   stringr_1.4.1   tools_4.2.1    
  utils_4.2.1     xfun_0.33       yaml_2.3.5 
```

**Comment by naikymen** [comment]
After pasting the image everywhere I found out where `include_graphics` ends up looking for the image at the time of the error.

The only path that made Knit work was the notebook's directory: `R/Rmd1/results/image_1.png`.

I don't understand how to tell it to look for the image at the project's root directory: `results/image_1.png`

**Comment by naikymen** [comment]
Well, I've found a workaround.

Here is an updated example: [test_2.zip](https://github.com/yihui/knitr/files/9592212/test_2.zip)

* `notebook1.Rmd` will now Knit and inlcude the image (using the normalized path), because I set `error=F`.
* `notebook1_error.Rmd` is identical, except that `error=T` (the default for `include_graphics`) and will not knit.

It seems to me that the error is raised mistakenly by `include_graphics`, because omitting the check results in the correct output.

 Is this correct?

**Comment by naikymen** [comment]
Possible culpirt:

https://github.com/yihui/knitr/blob/6bfffe9c1ae8c84f0f2ae07cd3c4b00876757587/R/plot.R#L492

**Comment by jameelalsalam** [comment]
I have also been struggling with and trying to figure out this issue. I believe that part of the issue is that if you pass an absolute path to knitr::include_graphics that it will try to convert it to a relative path, but it converts it to a path relative to the Rmd file, not relative to the output directory as is should.

This in turn is because `knitr::opts_knit$get("output.dir")` returns the Rmd file location, not the output directory. This is used inside of include_graphics.

Reprex here:

``` r
root_dir <- tempdir()

knitr_output_dir <- "to receive path from knit"

fs::dir_create(root_dir, "rmd-dir")
cat(
"---\ntitle: test\noutput: html_document\n---\n\n```{r}\n

# only important part of Rmd file:
knitr_output_dir <<- knitr::opts_knit$get('output.dir')\n```",
file = fs::path(root_dir, "rmd-dir/test.Rmd"))

fs::dir_create(root_dir, "output")
rmarkdown::render(
  fs::path(root_dir, "rmd-dir/test.Rmd"),
  output_dir = fs::path(root_dir, "output")
)
#> processing file: test.Rmd
#> output file: test.knit.md
#> "C:/Program Files/RStudio/bin/quarto/bin/tools/pandoc" +RTS -K512m -RTS test.knit.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output pandoc4d2879d05b93.html --lua-filter "C:\Users\jalsal02\R\win-library\4.1\rmarkdown\rmarkdown\lua\pagebreak.lua" --lua-filter "C:\Users\jalsal02\R\win-library\4.1\rmarkdown\rmarkdown\lua\latex-div.lua" --self-contained --variable bs3=TRUE --section-divs --template "C:\Users\jalsal02\R\win-library\4.1\rmarkdown\rmd\h\default.html" --no-highlight --variable highlightjs=1 --variable theme=bootstrap --mathjax --variable "mathjax-url=https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML" --include-in-header "C:\Users\jalsal02\AppData\Local\Temp\RtmpklzUsb\rmarkdown-str4d283a383e3d.html"
#> 
#> Output created: C:/Users/jalsal02/AppData/Local/Temp/RtmpklzUsb/output/test.html

print(knitr_output_dir)
#> [1] "C:/Users/jalsal02/AppData/Local/Temp/RtmpklzUsb/rmd-dir"
```

<sup>Created on 2022-09-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> - Session info ---------------------------------------------------------------
#>  setting  value
#>  version  R version 4.1.3 (2022-03-10)
#>  os       Windows 10 x64 (build 19042)
#>  system   x86_64, mingw32
#>  ui       RTerm
#>  language (EN)
#>  collate  English_United States.1252
#>  ctype    English_United States.1252
#>  tz       America/New_York
#>  date     2022-09-21
#>  pandoc   2.18 @ C:/Program Files/RStudio/bin/quarto/bin/tools/ (via rmarkdown)
#> 
#> - Packages -------------------------------------------------------------------
#>  package     * version date (UTC) lib source
#>  bslib         0.3.1   2021-10-06 [1] CRAN (R 4.1.1)
#>  cli           3.3.0   2022-04-25 [1] CRAN (R 4.1.3)
#>  digest        0.6.29  2021-12-01 [1] CRAN (R 4.1.3)
#>  evaluate      0.15    2022-02-18 [1] CRAN (R 4.1.3)
#>  fastmap       1.1.0   2021-01-25 [1] CRAN (R 4.1.1)
#>  fs            1.5.2   2021-12-08 [1] CRAN (R 4.1.3)
#>  glue          1.6.2   2022-02-24 [1] CRAN (R 4.1.3)
#>  highr         0.9     2021-04-16 [1] CRAN (R 4.1.1)
#>  htmltools     0.5.2   2021-08-25 [1] CRAN (R 4.1.1)
#>  jquerylib     0.1.4   2021-04-26 [1] CRAN (R 4.1.1)
#>  jsonlite      1.8.0   2022-02-22 [1] CRAN (R 4.1.3)
#>  knitr         1.40    2022-08-24 [1] CRAN (R 4.1.3)
#>  magrittr      2.0.3   2022-03-30 [1] CRAN (R 4.1.3)
#>  R6            2.5.1   2021-08-19 [1] CRAN (R 4.1.1)
#>  reprex        2.0.1   2021-08-05 [1] CRAN (R 4.1.1)
#>  rlang         1.0.5   2022-08-31 [1] CRAN (R 4.1.3)
#>  rmarkdown     2.16    2022-08-24 [1] CRAN (R 4.1.3)
#>  rstudioapi    0.13    2020-11-12 [1] CRAN (R 4.1.1)
#>  sass          0.4.1   2022-03-23 [1] CRAN (R 4.1.3)
#>  sessioninfo   1.2.2   2021-12-06 [1] CRAN (R 4.1.3)
#>  stringi       1.7.6   2021-11-29 [1] CRAN (R 4.1.2)
#>  stringr       1.4.0   2019-02-10 [1] CRAN (R 4.1.1)
#>  withr         2.5.0   2022-03-03 [1] CRAN (R 4.1.3)
#>  xfun          0.33    2022-09-12 [1] CRAN (R 4.1.3)
#>  yaml          2.3.5   2022-02-21 [1] CRAN (R 4.1.2)
#> 
#>  [1] C:/Users/jalsal02/R/win-library/4.1
#>  [2] C:/Program Files/R/R-4.1.3/library
#> 
#> ------------------------------------------------------------------------------
```

</details>

**Comment by jameelalsalam** [comment]
@naikymen Another approach you should be aware of is to pass absolute path to include_graphics (e.g., what your first example did with here::here()), and then `include_graphics(file, rel_path = FALSE)` to suppress the behavior to transform absolute into relative path.

**Comment by yihui** [comment]
Fixed in dev versions. Two separate problems were tangled here:

**1. The false error that blocked your knit.** `include_graphics()` converts an absolute path to a relative one (relative to knitr's working dir), then checks the file exists. Because you set `root.dir = here::here()`, the existence check ran from the project root while the path had been made relative to the input dir (`R/Rmd1/`) — mismatch → `Cannot find the file(s)` → knit aborted. The check now uses the original absolute path, so it no longer produces this false negative. This is why `error = FALSE` gave you correct output: only the check was wrong, not the emitted path.

Fixed with dev knitr alone:

```r
remotes::install_github('yihui/knitr')
```

**2. The wrong base directory (the deeper issue @cderv described).** When converting absolute → relative, knitr used its working directory (the *input* dir) as the base, not the directory of the *final rendered document*. When those differ in depth (e.g. rendering to a separate `output_dir`, or pkgdown → `./docs`, r-lib/pkgdown#2334), the resulting path is wrong. knitr now prefers the output directory when rmarkdown communicates it. This needs **both** dev packages:

```r
remotes::install_github('yihui/knitr')
remotes::install_github('rstudio/rmarkdown')
```

In your specific example the input and output dirs are both two levels deep, so `../../results/image_1.png` happened to resolve correctly from both — dev knitr alone is enough for you. `rel_path = FALSE` remains a good option when you know absolute paths work, as @jameelalsalam noted.
